### PR TITLE
Simplify data structures and remove redundant interfaces

### PR DESCRIPTION
**Pull request by chrisdoherty4**
**Summary**

The existing machine reconciliation code leveraged 3 different data structures (`TinkerbellMachineReconciler`, `machineReconcileContext` and `baseMachineReconcileContext`) and 2 interfaces (`ReconcileContext` and `BaseMachineReconcileContext`). 

The `machineReconcileContext` extends the `baseMachineReconcileContext` but the relationship is misappropriated and superfluous. 

The `ReconcileContext` and the `BaseMachineReconcileContext` have no discernible purpose.

**Changes**
* Delete the `ReconcileContext` and `BaseMachineReconcileContext` interfaces.
* Combine the `baseMachineReconcileContext` and `machineReconcileContext` into a single data structure, `machineReconcileScope`. 
* Move `machineReconcileScope` construction logic to the `TinkerbellMachineReconciler.Reconcile()` call keeping the construction in a single place.

**Motivation**
* Simpler code is easier to maintain.
* Updating to newer versions of Cluster API are easier when we understand the code better.

**What's next?**
* There's too close a relationship between the construction logic and delete reconciliation.
* Rework package structure so objects can be better tested (probably through the use of internal packages).
* Evaluating other logic to check for bugs. On performing this refactor there were some interesting smells that could contribute to recent reports of objects not reconciling.
* Updating to a recent Cluster API version.

### CODE REVIEW

File: controllers/machine_reconcile_scope.go
Comment (chrisdoherty4) [inline controllers/machine_reconcile_scope.go:1]: The functionality in this file is 90% copy-and-paste from the previous machine.go and base.go files. Changes in functionality have been highlighted with comments.

File: controllers/machine_reconcile_scope.go [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+
+type machineReconcileScope struct {
+	//*baseMachineReconcileContext
+	log               logr.Logger
+	ctx               context.Context
+	tinkerbellMachine *infrastructurev1.TinkerbellMachine
+	patchHelper       *patch.Helper
+	client            client.Client
+
+	machine              *clusterv1.Machine
```
Comment (chrisdoherty4) [inline controllers/machine_reconcile_scope.go:90]: These 3 fields were moved from the base object to ensure the existing functions could work without alteration.

File: controllers/tinkerbellmachine_controller.go [hunk 2]
```diff
@@ -54,30 +56,79 @@ type TinkerbellMachineReconciler struct {
 // +kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=jobs,verbs=get;list;watch;create
 
 // Reconcile ensures that all Tinkerbell machines are aligned with a given spec.
-func (tmr *TinkerbellMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	bmrc, err := tmr.newReconcileContext(ctx, req.NamespacedName)
```
Comment (chrisdoherty4) [inline controllers/tinkerbellmachine_controller.go:58]: The implementation of the `newReconcileContext()` has been moved to the `Reconcile()` call (lines 64-91).

File: controllers/tinkerbellmachine_controller.go [hunk 3]
```diff
@@ -54,30 +56,79 @@ type TinkerbellMachineReconciler struct {
...
+	// Typically, this is something akin to cloud-init user-data.
+	bootstrapCloudConfig, err := scope.getReadyBootstrapCloudConfig(machine)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("receiving bootstrap cloud config: %w", err)
+	}
+	if bootstrapCloudConfig == "" {
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
-	mrc, err := bmrc.IntoMachineReconcileContext()
```
Comment (chrisdoherty4) [inline controllers/tinkerbellmachine_controller.go:71]: The implementation of `IntoMachineReconcileContext()` has been moved to the `Reconcile()` call (lines 97-131).

File: controllers/tinkerbellmachine_controller_test.go [hunk 4]
```diff
@@ -596,8 +596,8 @@ func Test_Machine_reconciliation(t *testing.T) {
 	t.Run("fails_when", func(t *testing.T) {
 		t.Parallel()
 
-		t.Run("reconciler_is_nil", machineReconciliationFailsWhenReconcilerIsNil)                     //nolint:paralleltest
-		t.Run("reconciler_has_no_client_set", machineReconciliationFailsWhenReconcilerHasNoClientSet) //nolint:paralleltest
+		t.Run("reconciler_is_nil", machineReconciliationPanicsWhenReconcilerIsNil)                     //nolint:paralleltest
+		t.Run("reconciler_has_no_client_set", machineReconciliationPanicsWhenReconcilerHasNoClientSet) //nolint:paralleltest
```
Comment (chrisdoherty4) [inline controllers/tinkerbellmachine_controller_test.go:600]: See inline comments for explanation on the panic.

**Review by jacobweinstock — APPROVED** [review]

**Comment by chrisdoherty4** [comment]
@mergifyio queue

**Pull request merged**